File: LeetCode/C/103-tree-二叉树的锯齿形层次遍历.cpp
/*
 * @lc app=leetcode.cn id=103 lang=cpp
 *
 * [103] 二叉树的锯齿形层次遍历
 */
#include "main.h"
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
  public:
    vector<vector<int>> zigzagLevelOrder(TreeNode *root) {
        if (root == nullptr)
            return {};
        vector<vector<int>> res;
        dfs(root, 0, res);
        for (int i = 0; i < res.size(); i++)
            if (i % 2 == 1)
                reverse(res[i].begin(), res[i].end());
        return res;
    }

    void dfs(TreeNode *root, int level, vector<vector<int>> &res) {
        if (level == res.size())
            res.push_back({});
        res[level].push_back(root->val);
        if (root->left)
            dfs(root->left, level + 1, res);
        if (root->right)
            dfs(root->right, level + 1, res);
    }
};
// @lc code=end

File: LeetCode/Go/103-tree-二叉树的锯齿形层次遍历.go
/*
 * @lc app=leetcode.cn id=103 lang=golang
 *
 * [103] 二叉树的锯齿形层次遍历
 */
package leetcode

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	var res = [][]int{}

	var levelOrder func(*TreeNode, int)
	levelOrder = func(root *TreeNode, level int) {
		if root == nil {
			return
		}
		if level >= len(res) {
			res = append(res, []int{})
		}
		res[level] = append(res[level], root.Val)
		levelOrder(root.Left, level + 1)
		levelOrder(root.Right, level + 1)
	}

	levelOrder(root, 0)	
	for i := 0; i < len(res); i++ {
		if i & 1 == 1 {
			var l, r = 0, len(res[i]) - 1
			for l < r {
				res[i][l], res[i][r] = res[i][r], res[i][l]
				l++
				r--
			}
		}
	}
	return res
}
// @lc code=end

